relay-common/src/metrics.rs

```diff
@@ -64,30 +65,96 @@ where
     })
 }
 
+/// Any type that implements this trait can be set as a timer
+/// A typical implementation would be something like this
+///
+/// ```ignore
+///
+///enum MyTimer {
+///    TimeSpentDoingA,
+///    TimeSpentDoingB,
+///}
+///
+///impl TimerMetric for MyTimer {
+///    fn name(&self) -> &str {
+///        match self {
+///            Self::TimeSpentDoingA => "processA.millisecs",
+///            Self::TimeSpentDoingB => "processB.millisecs",
+///        }
+///    }
+///}
+/// ```
+///
+/// The type can then be used with any of the metric timer macros.
+///
+/// ```ignore
+/// let start_time  = std::time::Instant::now();
+/// metric!(timer(MyTimer::TimeSpentDoingA) = start_time.elapsed());
+/// ```
+pub trait TimerMetric {
+    /// Returns the timer metric name that will be sent to statsd (DataDog or whatever
+    /// collection server you use)
+    fn name(&self) -> Cow<'_, str>;
```

Comment:
To simplify the interface, how about `&str` here? I think we could even go further and enforce `&'static str`.

Reply:
changed to 'static str ( had to refactor metric `event.protocol.v<protocol_version>` to
`event.protocol` with version set as tag

Reply:
Ok